Device API and Specification Verification during Registration

twoshark opened the pull request:
### EvaluateSpecificationResponse
- consider http response before attempting to decode device spec

### VerifyDeviceAPI
- add method to ensure device adheres to DC api spec
- hydrates spec information into device

### Specification Check during Registration
- attempt to request and load spec from device after initial health check has passed
- do not activate discovered devices if init, health or spec pull fail
- delay `.Active = true` until spec has successfully been loaded
- add error logging for registration failures

### Local Dev Improvement
- Add db refresh option to make target & script

twoshark committed:
Endpoint Response Type
add ResponseType to Endpoint object and related params
add valid ResponseType to endpoints repo test

twoshark committed:
Update EvaluateSpecificationResponse
consider http response before attempting to decode device spec

twoshark committed:
Add DB Refresh Option to `make test` for Local Dev
accept `REFRESH` var in Makefile and pass value to `test.sh`
shut down docker container, delete volumes, restart and rerun migrations if REFRESH=true

twoshark committed:
Specification Check during Registration
attempt to request and load spec from device after initial health check has passed
break / do not activate discovered devices if init, health or spec pull fail
delay `.Active = true` until spec has successfully been loaded

twoshark committed:
Lint
add method to test and and activate newly discovered devices

twoshark committed:
Remove superflous Endpoint.ResponseType
this is handled by Endpoint.Parameter[].Type

dtp263 commented on src/device/specification.go:
```diff
@@ -0,0 +1,25 @@
+	return nil
+}
+
+func (d *Device) LoadFromSpecification(spec Device) {
```
pointer methods dont get mocked which make it harder to test down the line. Id recommend returning itself is a more flexible move. it also then allows for chaining of functions.

```
# before
func (d *Device) LoadFromSpecification(spec Device) {

# after
func (d Device) LoadFromSpecification(spec Device) Device {

dtp263 commented on src/device/specification.go:
```diff
@@ -0,0 +1,25 @@
+
+import "fmt"
+
+func (d Device) RequestSpecification(client Client) error {
```
test?

twoshark replied:
Added!

dtp263 commented on src/cluster/registration.go:
```diff
@@ -33,19 +33,39 @@ func (c DeviceCluster) DeviceDiscovery(scanDurationSeconds int) {
 	}
 }
 
+func (c DeviceCluster) initNewDevice(tmpNewDevice device.NewDeviceParams) error {
```
this function does a bunch of stuff. it prolly deserves some in-comments/docs

twoshark committed:
Remove LoadFromSpecification from * device
update method to return a Device for mock-ability

twoshark committed:
Refactor initNewDevice
add api verification steps to a new method, verifyDeviceAPI
add comments
bubble up device through verification and init methods to allow for hydration of device info

twoshark committed:
Refactor Discovery Flow Update
Remove extra layers of methods

twoshark committed:
Generate Mocks

twoshark committed:
Init Logger for Device FromNewDevice

twoshark committed:
Update Device Repository Test
Specifically test expected device properties so things like the logger can be ignored

twoshark committed:
Test VerifyDeviceAPI
Update Device interface to include VerifyDeviceAPI
Add VerifyDeviceAPI test to suite
Update warnings for Specification pull
frame Specification test file

twoshark committed:
Specification Tests
add test for LoadFromSpecification
add test for RequestSpecification

twoshark committed:
Specification Tests pt 2
add test for LoadFromSpecification
add test for RequestSpecification

twoshark committed:
Lint
add ListenForDevices to reduce DeviceDiscovery method length

twoshark committed:
Cleanup Comments

dtp263 commented on src/device/specification.go:
```diff
@@ -0,0 +1,26 @@
+	d.Description = spec.Description
+	d.Endpoints = spec.Endpoints
+	d.Name = spec.Name
+	return d
```
i think we are missing some assignments here. most importantly `Host` and `Port`. Setting these will enable us to track the same device as it changes IPs and maintains the same MAC address.

dtp263 commented on src/device/specification.go:
```diff
@@ -0,0 +1,26 @@
+	return spec, nil
+}
+
+func (d Device) LoadFromSpecification(spec Device) Device {
```
also i suppose we can skip for now. but i think this is also where `Endpoints` would be loaded as well.

twoshark commented on scripts/test.sh:
```diff
@@ -1,8 +1,29 @@
+
+    # Clear old volumes
+    docker volume rm devicecommander_postgres || true
+    docker volume rm devicecommander_pgadmin || true
```
This was I think meant as more of a dev convenience/break glass procedure. The default is not to tear anything down.

I don't feel strongly about it so I'm happy to remove the option.

dtp263 replied:
Meh it's prolly fine. If it helps you dev 🤷🏼‍♂️

twoshark committed:
Update src/postgres/device/device_repository_test.go

Co-authored-by: Olin Davis <olin.davis20@gmail.com>

twoshark committed:
Add helper `WrapError` to utilities

twoshark committed:
Merge branch 'specification-eval' of https://github.com/RNA-VT/DeviceCommander into specification-eval

twoshark committed:
WIP fixes
- Updated Mocks
- Add Error constants and use in tests
- gofmt

twoshark committed:
Updates
update go.mod
docker-compose -> docker compose
